File: src/services/medicineService.js
// ==========================================================
// medicineService.js
// Service xử lý API quản lý thuốc / lịch uống thuốc
// Tương tác trực tiếp với Backend Flask API qua api.js (Axios)
// ==========================================================

import api from "./api";

/**
 * Service quản lý đầy đủ các thao tác CRUD và truy vấn danh sách thuốc từ Backend.
 */
const medicineService = {
  /**
   * Lấy toàn bộ danh sách thuốc.
   * Endpoint: GET /api/medicines
   * @returns {Promise<Array>} Danh sách thuốc
   */
  async getAll() {
    try {
      const response = await api.get("/medicines", { params: { per_page: 100 } });
      return Array.isArray(response)
        ? response
        : Array.isArray(response?.data)
        ? response.data
        : response?.data?.items || response?.items || [];
    } catch (error) {
      console.error("Lỗi khi lấy danh sách thuốc:", error.message);
      throw error;
    }
  },

  /**
   * Lấy chi tiết thông tin thuốc theo ID.
   * Endpoint: GET /api/medicines/:id
   * @param {number|string} id ID thuốc
   * @returns {Promise<Object>} Chi tiết thuốc
   */
  async getById(id) {
    try {
      const response = await api.get(`/medicines/${id}`);
      return response.data || null;
    } catch (error) {
      console.error(`Lỗi khi lấy thông tin thuốc ID ${id}:`, error.message);
      throw error;
    }
  },

  /**
   * Thêm mới thuốc vào cơ sở dữ liệu.
   * Endpoint: POST /api/medicines
   * @param {Object} medicineData Dữ liệu thuốc mới
   * @returns {Promise<Object>} Thông tin thuốc đã được tạo
   */
  async create(medicineData) {
    try {
      const response = await api.post("/medicines", medicineData);
      return response.data;
    } catch (error) {
      console.error("Lỗi khi thêm mới thuốc:", error.message);
      throw error;
    }
  },

  /**
   * Cập nhật thông tin thuốc theo ID.
   * Endpoint: PUT /api/medicines/:id
   * @param {number|string} id ID thuốc cần cập nhật
   * @param {Object} medicineData Dữ liệu cập nhật
   * @returns {Promise<Object>} Thông tin thuốc sau khi cập nhật
   */
  async update(id, medicineData) {
    try {
      const response = await api.put(`/medicines/${id}`, medicineData);
      return response.data;
    } catch (error) {
      console.error(`Lỗi khi cập nhật thuốc ID ${id}:`, error.message);
      throw error;
    }
  },

  /**
   * Xóa thuốc theo ID khỏi cơ sở dữ liệu.
   * Endpoint: DELETE /api/medicines/:id
   * @param {number|string} id ID thuốc cần xóa
   * @returns {Promise<Object>} Kết quả phản hồi từ API
   */
  async delete(id) {
    try {
      const response = await api.delete(`/medicines/${id}`);
      return response;
    } catch (error) {
      console.error(`Lỗi khi xóa thuốc ID ${id}:`, error.message);
      throw error;
    }
  },

  /**
   * Tìm kiếm thuốc theo từ khóa tên hoặc liều lượng.
   * Endpoint: GET /api/medicines/search?keyword=...
   * @param {string} keyword Từ khóa tìm kiếm
   * @returns {Promise<Array>} Danh sách kết quả tìm kiếm
   */
  async search(keyword = "") {
    try {
      const response = await api.get("/medicines/search", {
        params: { keyword, per_page: 100 },
      });
      return Array.isArray(response)
        ? response
        : Array.isArray(response?.data)
        ? response.data
        : response?.data?.items || response?.items || [];
    } catch (error) {
      console.error("Lỗi khi tìm kiếm thuốc:", error.message);
      throw error;
    }
  },

  /**
   * Lấy danh sách các thuốc sắp hết (số lượng dưới ngưỡng).
   * Endpoint: GET /api/medicines/low-stock
   * @returns {Promise<Array>} Danh sách thuốc sắp hết
   */
  async getLowStock() {
    try {
      const response = await api.get("/medicines/low-stock");
      return Array.isArray(response)
        ? response
        : Array.isArray(response?.data)
        ? response.data
        : response?.data?.items || response?.items || [];
    } catch (error) {
      console.error("Lỗi khi lấy danh sách thuốc sắp hết:", error.message);
      throw error;
    }
  },

  /**
   * Lấy danh sách các thuốc đã hết hạn sử dụng.
   * Endpoint: GET /api/medicines/expired
   * @returns {Promise<Array>} Danh sách thuốc hết hạn
   */
  async getExpired() {
    try {
      const response = await api.get("/medicines/expired");
      return Array.isArray(response)
        ? response
        : Array.isArray(response?.data)
        ? response.data
        : response?.data?.items || response?.items || [];
    } catch (error) {
      console.error("Lỗi khi lấy danh sách thuốc hết hạn:", error.message);
      throw error;
    }
  },
};

export const getMedicines = () => [];

export default medicineService;

File: services/medicine_service.py
# ==========================================================
# medicine_service.py
# Tầng Service quản lý danh mục thuốc (Medicine)
# Chịu trách nhiệm thực hiện nghiệp vụ, phân trang, lọc, sắp xếp và thao tác DB
# Khớp 100% với Model Medicine và bảng MySQL `medicines`
# ==========================================================

# ==========================================================
# Import thư viện
# ==========================================================

import logging
from typing import Any, Dict, List, Optional, Tuple

from sqlalchemy import asc, desc, or_

from database import db
from models.medicine import Medicine

# ==========================================================
# Cấu hình Logger
# ==========================================================

logger = logging.getLogger(__name__)


# ==========================================================
# Class MedicineService
# ==========================================================

class MedicineService:
    """
    Lớp dịch vụ xử lý nghiệp vụ liên quan đến danh mục thuốc.
    """

    # ======================================================
    # Lấy danh sách thuốc (Phân trang, Tìm kiếm, Sắp xếp)
    # ======================================================
    @staticmethod
    def get_all(
        page: int = 1,
        per_page: int = 10,
        search: Optional[str] = None,
        sort_by: str = "medicine_id",
        order: str = "asc"
    ) -> Tuple[List[Medicine], int]:
        """
        Lấy danh sách thuốc có hỗ trợ Phân trang, Tìm kiếm và Sắp xếp.

        Args:
            page (int): Số trang hiện tại.
            per_page (int): Số lượng bản ghi mỗi trang.
            search (Optional[str]): Từ khóa tìm kiếm (Tên thuốc, Mã thuốc, Liều lượng).
            sort_by (str): Tên cột sắp xếp.
            order (str): Thứ tự sắp xếp ('asc' hoặc 'desc').

        Returns:
            Tuple[List[Medicine], int]: (Danh sách Medicine, Tổng số bản ghi)
        """
        logger.info("Service: Truy vấn danh sách thuốc (Page=%s, PerPage=%s)", page, per_page)
        query = Medicine.query

        if search:
            search_pattern = f"%{search}%"
            query = query.filter(
                or_(
                    Medicine.medicine_id.like(search_pattern),
                    Medicine.medicine_name.like(search_pattern),
                    Medicine.dosage.like(search_pattern),
                    Medicine.frequency.like(search_pattern)
                )
            )

        total = query.count()

        sort_column = getattr(Medicine, sort_by, Medicine.medicine_id)
        if order.lower() == "desc":
            query = query.order_by(desc(sort_column))
        else:
            query = query.order_by(asc(sort_column))

        medicines = query.offset((page - 1) * per_page).limit(per_page).all()
        return medicines, total

    # ======================================================
    # Lấy thuốc theo medicine_id
    # ======================================================
    @staticmethod
    def get_by_id(medicine_id: str) -> Optional[Medicine]:
        """
        Lấy thông tin chi tiết một loại thuốc theo medicine_id.

        Args:
            medicine_id (str): Mã thuốc.

        Returns:
            Optional[Medicine]: Đối tượng Medicine hoặc None.
        """
        logger.info("Service: Truy vấn thuốc theo ID: '%s'", medicine_id)
        return db.session.get(Medicine, medicine_id)

    # ======================================================
    # Thêm mới thuốc
    # ======================================================
    @staticmethod
    def create(data: Dict[str, Any]) -> Medicine:
        """
        Thêm mới một loại thuốc vào cơ sở dữ liệu MySQL.

        Args:
            data (Dict[str, Any]): Dữ liệu thuốc.

        Returns:
            Medicine: Đối tượng Medicine vừa tạo.

        Raises:
            Exception: Lỗi lưu cơ sở dữ liệu.
        """
        logger.info("Service: Tiến hành tạo thuốc mới ID: '%s'", data.get("medicine_id"))
        try:
            medicine = Medicine(
                medicine_id=data["medicine_id"],
                medicine_name=data.get("medicine_name"),
                dosage=data.get("dosage"),
                frequency=data.get("frequency"),
            )

            db.session.add(medicine)
            db.session.commit()
            logger.info("Service: Tạo thành công thuốc ID: '%s'", medicine.medicine_id)
            return medicine
        except Exception as error:
            db.session.rollback()
            logger.error("Service: Lỗi khi thêm mới thuốc: %s", str(error))
            raise error

    # ======================================================
    # Cập nhật thông tin thuốc
    # ======================================================
    @staticmethod
    def update(medicine_id: str, data: Dict[str, Any]) -> Optional[Medicine]:
        """
        Cập nhật thông tin thuốc theo medicine_id.

        Args:
            medicine_id (str): Mã thuốc cần cập nhật.
            data (Dict[str, Any]): Dữ liệu mới.

        Returns:
            Optional[Medicine]: Đối tượng Medicine sau cập nhật hoặc None.

        Raises:
            Exception: Lỗi cập nhật cơ sở dữ liệu.
        """
        logger.info("Service: Tiến hành cập nhật thuốc ID: '%s'", medicine_id)
        try:
            medicine = db.session.get(Medicine, medicine_id)
            if medicine is None:
                logger.warning("Service: Không tìm thấy thuốc ID: '%s'", medicine_id)
                return None

            if "medicine_name" in data:
                medicine.medicine_name = data["medicine_name"]
            if "dosage" in data:
                medicine.dosage = data["dosage"]
            if "frequency" in data:
                medicine.frequency = data["frequency"]

            db.session.commit()
            logger.info("Service: Cập nhật thành công thuốc ID: '%s'", medicine_id)
            return medicine
        except Exception as error:
            db.session.rollback()
            logger.error("Service: Lỗi khi cập nhật thuốc ID '%s': %s", medicine_id, str(error))
            raise error

    # ======================================================
    # Xóa thuốc khỏi hệ thống
    # ======================================================
    @staticmethod
    def delete(medicine_id: str) -> bool:
        """
        Xóa một loại thuốc theo medicine_id.

        Args:
            medicine_id (str): Mã thuốc.

        Returns:
            bool: True nếu xóa thành công, False nếu không tìm thấy.

        Raises:
            Exception: Lỗi xóa cơ sở dữ liệu.
        """
        logger.info("Service: Tiến hành xóa thuốc ID: '%s'", medicine_id)
        try:
            medicine = db.session.get(Medicine, medicine_id)
            if medicine is None:
                logger.warning("Service: Không tìm thấy thuốc ID: '%s' để xóa", medicine_id)
                return False

            db.session.delete(medicine)
            db.session.commit()
            logger.info("Service: Xóa thành công thuốc ID: '%s'", medicine_id)
            return True
        except Exception as error:
            db.session.rollback()
            logger.error("Service: Lỗi khi xóa thuốc ID '%s': %s", medicine_id, str(error))
            raise error